File: Python/Problems/Problems/problem2587.py
import heapq
import math
import sys

import numpy as np

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem2587(Problem):
    def test(self) -> bool:
        nums = [2, -1, 0, 1, -3, 3, -3]

        expected = 6

        result = self.maxScore(nums)

        print(result)

        return result == expected

    def maxScore(self, nums: List[int]) -> int:
        n = len(nums)
        nums.sort()
        pre = 0
        for i in range(n - 1, -1, -1):
            pre += nums[i]
            if pre <= 0:
                return n - i - 1

        return n

File: C++/Problems/Problems/problem_2587.cpp
#include "problem_2587.h"

bool problem_2587::test() {
    std::vector nums  { 2, -1, 0, 1, -3, 3, -3 };

    int expected = 6;

    auto result = maxScore(nums);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_2587::maxScore(std::vector<int> &nums) {
    int n = nums.size();
    std::sort(nums.begin(), nums.end());
    long long pre = 0;
    for (int i = n - 1; i >= 0; --i)
    {
        pre += nums[i];
        if (pre <= 0)
            return n - i - 1;
    }

    return n;
}
